src/main/java/com/coreos/jetcd/EtcdWatchImpl.java

```diff
@@ -0,0 +1,294 @@
+package com.coreos.jetcd;
+
+import com.coreos.jetcd.api.*;
+import com.coreos.jetcd.options.WatchOption;
+import com.coreos.jetcd.util.ListenableSetFuture;
+import com.coreos.jetcd.watch.Watcher;
+import com.google.protobuf.ByteString;
+import io.grpc.stub.StreamObserver;
+import javafx.util.Pair;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+/**
+ * etcd watcher Implementation
+ */
+public class EtcdWatchImpl implements EtcdWatch {
+
+    private StreamObserver<WatchRequest> requestStream;
```

Comment:
generally, I think `volatile` here is enough, I don't want to make this giantic complicated. Suppliers.memorize is not that common used.